File: examples/Java/mdcliapi2.java
/**
 * (c) 2011 Arkadiusz Orzechowski
 *
 * This file is part of ZGuide
 *
 * ZGuide is free software; you can redistribute it and/or modify it under
 * the terms of the Lesser GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * ZGuide is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * Lesser GNU General Public License for more details.
 *
 * You should have received a copy of the Lesser GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
import java.util.Formatter;

import org.zeromq.ZContext;
import org.zeromq.ZFrame;
import org.zeromq.ZMQ;
import org.zeromq.ZMsg;

/**
 * Majordomo Protocol Client API, asynchronous Java version. Implements the
 * MDP/Worker spec at http://rfc.zeromq.org/spec:7.
 * 
 * @author Arkadiusz Orzechowski <aorzecho@gmail.com>
 */
public class mdcliapi2 {

    private String broker;
    private ZContext ctx;
    private ZMQ.Socket client;
    private long timeout = 2500;
    private boolean verbose;
    private Formatter log = new Formatter(System.out);

    public long getTimeout() {
        return timeout;
    }

    public void setTimeout(long timeout) {
        this.timeout = timeout;
    }

    public mdcliapi2(String broker, boolean verbose) {
        this.broker = broker;
        this.verbose = verbose;
        ctx = new ZContext();
        reconnectToBroker();
    }

    /**
     * Connect or reconnect to broker
     */
    void reconnectToBroker() {
        if (client != null) {
            ctx.destroySocket(client);
        }
        client = ctx.createSocket(ZMQ.DEALER);
        client.connect(broker);
        if (verbose)
            log.format("I: connecting to broker at %s...\n", broker);
    }

    /**
     * Returns the reply message or NULL if there was no reply. Does not attempt
     * to recover from a broker failure, this is not possible without storing
     * all unanswered requests and resending them all…
     */
    public ZMsg recv() {
        ZMsg reply = null;

        // Poll socket for a reply, with timeout
        ZMQ.Poller items = ctx.getContext().poller();
        items.register(client, ZMQ.Poller.POLLIN);
        if (items.poll(timeout * 1000) == -1)
            return null; // Interrupted

        if (items.pollin(0)) {
            ZMsg msg = ZMsg.recvMsg(client);
            if (verbose) {
                log.format("I: received reply: \n");
                msg.dump(log.out());
            }
            // Don't try to handle errors, just assert noisily
            assert (msg.size() >= 4);

            ZFrame empty = msg.pop();
            assert (empty.getData().length == 0);
            empty.destroy();

            ZFrame header = msg.pop();
            assert (MDP.C_CLIENT.equals(header.toString()));
            header.destroy();

            ZFrame replyService = msg.pop();
            replyService.destroy();

            reply = msg;
        }
        return reply;
    }

    /**
     * Send request to broker and get reply by hook or crook Takes ownership of
     * request message and destroys it when sent.
     */
    public void send(String service, ZMsg request) {
        assert (request != null);

        // Prefix request with protocol frames
        // Frame 0: empty (REQ emulation)
        // Frame 1: "MDPCxy" (six bytes, MDP/Client x.y)
        // Frame 2: Service name (printable string)
        request.addFirst(service);
        request.addFirst(MDP.C_CLIENT.newFrame());
        request.addFirst("");
        if (verbose) {
            log.format("I: send request to '%s' service: \n", service);
            request.dump(log.out());
        }
        request.send(client);
    }

    public void destroy() {
        ctx.destroy();
    }
}

File: examples/PHP/mdcliapi2.php
<?php
/*  =====================================================================
 * mdcliapi2.c
 *
 * Majordomo Protocol Client API (async version)
 * Implements the MDP/Worker spec at http://rfc.zeromq.org/spec:7.
 *
 * ---------------------------------------------------------------------
 * Copyright (c) 1991-2011 iMatix Corporation <www.imatix.com>
 * Copyright other contributors as noted in the AUTHORS file.
 *
 * This file is part of the ZeroMQ Guide: http://zguide.zeromq.org
 *
 * This is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or (at
 * your option) any later version.
 *
 * This software is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this program. If not, see
 * <http://www.gnu.org/licenses/>.
 * =====================================================================
 * @author Ian Barber <ian(dot)barber(at)gmail(dot)com>
 */

include_once 'zmsg.php';
include_once 'mdp.php';

class MDCli
{
    //  Structure of our class
    //  We access these properties only via class methods
    private $broker;
    private $context;
    private $client;	//  Socket to broker
    private $verbose;	//  Print activity to stdout
    private $timeout;	//  Request timeout
    private $retries;	//  Request retries

    /**
     * Constructor
     *
     * @param string  $broker
     * @param boolean $verbose
     */
    public function __construct($broker, $verbose = false)
    {
        $this->broker = $broker;
        $this->context = new ZMQContext();
        $this->verbose = $verbose;
        $this->timeout = 2500;           //  msecs
        $this->connect_to_broker();
    }

    /**
     * Connect or reconnect to broker
     */
    protected function connect_to_broker()
    {
        if ($this->client) {
            unset($this->client);
        }
        $this->client = new ZMQSocket($this->context, ZMQ::SOCKET_DEALER);
        $this->client->setSockOpt(ZMQ::SOCKOPT_LINGER, 0);
        $this->client->connect($this->broker);
        if ($this->verbose) {
            printf("I: connecting to broker at %s...", $this->broker);
        }
    }

    /**
     * Set request timeout
     *
     * @param int $timeout (msecs)
     */
    public function set_timeout($timeout)
    {
        $this->timeout = $timeout;
    }

    /**
     * Send request to broker
     * Takes ownership of request message and destroys it when sent.
     *
     * @param string $service
     * @param Zmsg   $request
     */
    public function send($service, Zmsg $request)
    {
        //  Prefix request with protocol frames
        //  Frame 0: empty (REQ emulation)
        //  Frame 1: "MDPCxy" (six bytes, MDP/Client x.y)
        //  Frame 2: Service name (printable string)
        $request->push($service);
        $request->push(MDPC_CLIENT);
        $request->push("");
        if ($this->verbose) {
            printf ("I: send request to '%s' service: %s", $service, PHP_EOL);
            echo $request->__toString();
        }
        $request->set_socket($this->client)->send();
    }

    /**
     * Returns the reply message or NULL if there was no reply. Does not
     * attempt to recover from a broker failure, this is not possible
     * without storing all unanswered requests and resending them all...
     *
     */
    public function recv()
    {
        $read = $write = array();

        //  Poll socket for a reply, with timeout
        $poll = new ZMQPoll();
        $poll->add($this->client, ZMQ::POLL_IN);
        $events = $poll->poll($read, $write, $this->timeout);

        //  If we got a reply, process it
        if ($events) {
            $msg = new Zmsg($this->client);
            $msg->recv();
            if ($this->verbose) {
                echo "I: received reply:", $request->__toString(), PHP_EOL;
            }
            //  Don't try to handle errors, just assert noisily
            assert ($msg->parts() >= 4);

            $msg->pop(); // empty

            $header = $msg->pop();
            assert($header == MDPC_CLIENT);

            $reply_service = $msg->pop();

            return $msg; //  Success
        } else {
            echo "W: permanent error, abandoning request", PHP_EOL;

            return;	//  Give up
        }
    }
}
